File: 백준/Silver/5555. 반지/반지.py
import sys

input = sys.stdin.readline

'''
반지 (https://www.acmicpc.net/problem/5555)
'''

def solution(t, N, rings):
    answer = 0
    tlength = len(t)
    for ring in rings:
        length = len(ring)
        ring *= 2
        for start in range(length):
            string = ring[start : start + tlength]
            if string == t:
                answer += 1
                break

    return answer

t = input().rstrip()
N = int(input())
rings = [input().rstrip() for _ in range(N)]
print(solution(t, N, rings))

File: 백준/Silver/5555. 반지/반지.java
import java.io.BufferedReader;
import java.io.FileInputStream;
import java.io.IOException;
import java.io.InputStream;
import java.io.InputStreamReader;

/**
 * https://www.acmicpc.net/problem/1181
 */
public class Main {

    public static int solution(String delimiter, int N, String[] words) {
        int dLength = delimiter.length();
        int answer = 0;
        for (String word : words) {
            int length = word.length();
            word = word.repeat(2);
            for (int i = 0; i < length; i++) {
                if (word.substring(i, i + dLength).equals(delimiter)) {
                    answer++;
                    break;
                }
            }
        }

        return answer;
    }

    public static void main(String[] args) throws IOException {
        InputStream input = System.in;
        BufferedReader br = new BufferedReader(new InputStreamReader(input));

        String delimiter = br.readLine();
        int N = Integer.parseInt(br.readLine());
        String[] words = new String[N];
        for (int i = 0; i < N; i++) {
            words[i] = br.readLine();
        }
        System.out.println(solution(delimiter, N, words));
    }
}